test/test_pytree.py

```diff
@@ -60,35 +42,55 @@ def run_test_with_leaf(leaf):
         run_test_with_leaf(bool)
         run_test_with_leaf(torch.randn(3, 3))
 
-    def test_flatten_unflatten_list(self):
+    @parametrize(
+        "pytree_impl,gen_expected_fn",
+        [
+            (
+                py_pytree,
+                lambda lst: py_pytree.TreeSpec(
+                    list, None, [py_pytree.LeafSpec() for _ in lst]
+                ),
+            ),
+            (cxx_pytree, lambda lst: cxx_pytree.tree_structure([0] * len(lst))),
+        ],
+    )
```

Comment:
> It sounds like we should add a tree_structure API to Python pytree. This would further improve the testing, because we wouldn't need to specify the difference between input generation for Python vs C++ pytree (we can call tree_structure on both APIs).

These test functions will test both `tree_flatten` and `tree_unflatten`. I think we need to explicitly construct the expected treespec for testing for the Python pytree. Because `tree_structure` calls `tree_flatten` and that is what we need to test for.

Reply:
My problem right now is that the Python and C++ pytree testing logic is different here (and in a bunch of other places). They really shouldn't be. If the claim is that we should be manually constructing TreeSpec, then we should manually construct the TreeSpec in both Python and C++.